### PR TITLE
feat(bigquery): implement array functions

### PR DESCRIPTION
**Pull request by cpcloud**
Add array function support to BigQuery.

### CODE REVIEW

**Comment by cpcloud** [comment]
BigQuery array tests are passing:

```
❯ pytest -m bigquery ibis/backends/tests/test_array.py -n auto -q
bringing up nodes...
.xx....x.x..x.x.....x.x.xxxxx.........x....                                                                                                                                                    [100%]
29 passed, 14 xfailed in 95.71s (0:01:35)
```

**Comment by cpcloud** [comment]
Test passing after implementing `zip`:

```
❯ pytest -m bigquery ibis/backends/tests/test_array.py -n auto -q
bringing up nodes...
..x..........x.x.x.xx....xx.xx......x..xx..                                                                                                                                                    [100%]
30 passed, 13 xfailed in 81.49s (0:01:21)
```

File: ibis/backends/bigquery/compiler.py [hunk 1]
```diff
@@ -130,16 +130,18 @@ def _quote_identifier(self, name):
...
         types = schema.types
 
         raw_rows = []
         for row in op.data.to_frame().itertuples(index=False):
             raw_row = ", ".join(
-                f"{self._translate(lit)} AS {name}"
+                f"{self._translate(lit.op())} AS {name}"
                 for lit, name in zip(
-                    map(ops.Literal, row, types), map(self._quote_identifier, names)
+                    map(ibis.literal, row, types), map(self._quote_identifier, names)
```
Comment (kszucs) [inline ibis/backends/bigquery/compiler.py:144]: Why is this change necessary?

**Review by kszucs — APPROVED** [review]
Nice! This change by itself could worth a blogpost.

File: ibis/backends/bigquery/compiler.py (hunk 1 above)
Comment (cpcloud) [reply ibis/backends/bigquery/compiler.py:144]: Because `ops.Literal` doesn't normalize `list` inputs for array types. It's only accepts hashable sequences:

```
In [5]: import ibis

In [6]: ibis.literal(["a"])
Out[6]: ('a',)

In [7]: import ibis.expr.operations as ops

In [8]: ops.Literal(["a"], dtype="array<string>")
...
IbisTypeError: argument passes none of the following rules: ...
```

Comment (kszucs) [reply ibis/backends/bigquery/compiler.py:144]: Could you please try that out on the 7.x branch?

**Comment by cpcloud** [comment]
Tests passing after adding support for `map` and `filter` methods:

```
…/ibis on  bigquery-array-functions is 📦 v6.1.0 via 🐍 v3.10.12 via ❄️  impure (ibis-3.10.12-env) took 5s
❯ pytest -m bigquery ibis/backends/tests/test_array.py -n auto -q
bringing up nodes...
x.x.xxx.....x.x...xx................x.x.x....                                                                                                                                                  [100%]
33 passed, 12 xfailed in 82.31s (0:01:22)
```
